### ISSUE TITLE
Cordova / PhoneGap / Ionic Support

### ISSUE THREAD

**Issue by timanglade** [realm/realm-js#261]

**Comment by appden** [comment]
I just rebased and pushed my older WIP branch, `sk-cordova`, which pretty much completes this task. :smile:

**Comment by appden** [comment]
I mean for iOS. For Android, this is pretty much blocked on #262 being done.

**Comment by keithdmoore** [comment]
👍 for Ionic support.  Hopefully Ionic2!

**Comment by mlynch** [comment]
This would be awesome and well received in the community for sure. Curious why this wouldn't Just Work given the recent realm-js release?

**Comment by pjhartin** [comment]
@mlynch I would assume because for phonegap, an interface between the js library to the native hardware of the device would need to be developed to actually bring the benefit of realms speed. But then again, I have no idea.

**Comment by alazier** [comment]
The main issue with supporting phonegap/cordova at the moment is the need to support android. React Native runs on JavascriptCore on both iOS and android which is how we support both platforms. This is not the case for cordova/phonegap.

**Comment by Kratos-SA** [comment]
+1

**Comment by caiobalthazar** [comment]
+1

**Comment by douglasjunior** [comment]
It's a good option to create a cordova-plugin to use realm-java and realm-cocoa?

As in https://github.com/mapsplugin/cordova-plugin-googlemaps

**Comment by PierBover** [comment]
> The main issue with supporting phonegap/cordova at the moment is the need to support android

@alazier can you expand on that?

**Comment by alazier** [comment]
@PierBover - the way the Realm binding works is that we hook directly into the JS context for whatever environment we are supporting. React-native uses the same version of JavascriptCore on both ios and android which makes integration pretty straightforward. Cordovoa uses whatever the builtin JS interpreter is on the device it is running on, which makes things much more difficult. It might be possible to make things work by rolling a custom version of Cordova but this isn't something we have looked into. If you just want to support Cordova/iOS then it should be possible to get things working using the current codebase by getting the JSContext from a webview and passing it into the Realm initialization methods, but this is not something that we can currently support or maintain.

**Comment by PierBover** [comment]
Ok I see the problem.

Probably the best way to go for Cordova/Phonegap is making a native plugin (since you already have native iOS and Android libraries) with a JS wrapper API that mimics the one used in your React Native implementation.

**Comment by virtualprodigy** [comment]
My team really wants Realm for Cordova/Ionic. Any updates on if/when Cordova support would rollout?

**Comment by tiagoMaua** [comment]
+1

**Comment by caiobalthazar** [comment]
I believe a plugin approach such as the one described by @PierBover would be the easiest way to do it.

**Comment by kitfactory** [comment]
+1

**Comment by faraazc** [comment]
+1 ionic2 support is really wanted in the community. This will open doors for PWA as well.

**Comment by AresDev** [comment]
+1

**Comment by gozur** [comment]
+1

**Comment by croth53** [comment]
+1 IONIC2 support is wanted

**Comment by krasnokojiy** [comment]
+1

**Comment by konrin** [comment]
+1

**Comment by IvanAR** [comment]
+1

**Comment by llearn** [comment]
+1

**Comment by gkTim** [comment]
Any news on this?

**Comment by jruef** [comment]
+1

**Comment by istx25** [comment]
As far as I know, this is not something that we are actively working on right now.

**Comment by zuhrulumam** [comment]
+1

**Comment by danmatthews** [comment]
Damn, it's a shame that this isn't something that's actively being worked on - i think it's a big market potentially.

In the meantime, if you're not working on this, can someone close it and update it with some official news?

In the meantime guys - the best alternative is [PouchDB](pouchdb.com) with the [sqlite adapter](https://github.com/nolanlawson/pouchdb-adapter-cordova-sqlite).

**Comment by jruef** [comment]
@danmatthews yeah it's a shame. I also ended up using PouchDB with the sqlite adapter. I develop for react native so you can use [pouchdb-adapter-react-native-sqlite](https://github.com/noradaiko/pouchdb-adapter-react-native-sqlite) and [pouchdb-react-native](https://github.com/stockulus/pouchdb-react-native) :-)

@danmatthews by the way the first link in your comment is broken

**Comment by kristiandupont** [comment]
Yeah, sorry, the official news is that @istx25 is right. This is not something we are working on at the moment. I realize that it's a popular request and definitely something we will look into but at the moment, the JS team has plenty of work with the currently supported platforms. I'll keep the issue open as it's neither done nor rejected, just idle :-)

**Comment by danmatthews** [comment]
No worries! Thanks for getting back to us and letting us know the status.

**Comment by AsimNet** [comment]
+1

**Comment by bmunkholm** [comment]
https://github.com/airamrguez/cordova-plugin-realm is making an attempt.

**Comment by useitsoftware** [comment]
+1 any chance  of rethinking on this? angular2/angular4 and ionic 2 are growing platforms with a lot of popularity. You could partner with the ionic team to share the development burden.

**Comment by malhobayyeb** [comment]
+1

**Comment by tryadelion** [comment]
@kristiandupont  @alazier  could you please try to throw this back on the #Maybe deck ?  Ionic2/Angular2 & ionic3/Angular4 both lack a solid mautre & stable database system like Realm  behind and their platform and numbers of apps is growing fast. showcase.ionicframework.com/apps/newest

**Comment by nazrdogan** [comment]
+1

**Comment by dav021** [comment]
+1

**Comment by yosuke1985** [comment]
+1

**Comment by ozexpert** [comment]
What's the status? I've used React Native with Realm. But want to switch over using Ionic/Cordova with Realm. Blazingly fast database!!

**Comment by tryadelion** [comment]
What is the status again? it's been almost half a year since @kristiandupont  's statement. React native isn't the only cross platform JS framework, centering yourselves on it forces the rest of us to look elsewhere for worse solutions.

**Comment by syc-pad** [comment]
+1

**Comment by tryadelion** [comment]
Feature Request:

## Goals

To achieve realm connection between a  cordova-based hybrid app  and the native library code by creating a cordova  Typescript plugin that takes a generically Inputted syntax (RealmJS syntax?, swift syntax?) that redirects to the desired methods in the native code.

Examples of how this works:

https://github.com/arnesson/cordova-plugin-firebase

## Expected Results

TS:

```
import { RealmDB, RealmObject} from "Realm";
private car : RealmObject = {brand: "honda",subtype:"civic",codename:"grizzlybear",year:"2017",id: 12043};
constructor(realmInstance : RealmDB){
    realmInstance.open().subscribe( () => {
        realmInstance.write(car).subscribe( (res : Boolean) => {
          console.log(res);
        }
    });
}

```
Native IOS:

_let realm = try! Realm()

etc...

native Android:_

_new Realm()

etc,,,,_

## REASON

Realm.JS isn't made to be used as front end for webview-based hybrid apps, so there must be a way to use it. the easiest is to move the code "behind", and connect to the desired core when platform meets needs.

## Possible compatible versions
Android
Ios
Windows 10 UWP  (surface, desktop, windows phone All in One)

**Comment by bmunkholm** [comment]
We currently don't have the spare capacity to take on more new platforms. There is a significant amount of work involved in creating, maintaining and providing support for each platform. So we would need the community to step up and assist with this effort. I can see that some effort is being put into https://github.com/airamrguez/cordova-plugin-realm.

**Comment by superandrew** [comment]
+1

**Comment by tryadelion** [comment]
@bmunkholm  hey brian, we understand that the capacity to make a new platform from scratch can be breathtaking. the issue with cordova-plugin-realm is that it relies on premade and predefined scheme files, built over json. 
I have no experience on React Native, but i suppose that what it does is detect concrete javascript method calls, and direct them to a native code that takes care of it. correct me if i'm wrong.
doing a Cordova wrapper that connected a Realm-JS'subset' of TS/JS calls with the already existant Realm native calls (and the already existing RealmJS syntax and boilerplate) could take some time to tinker, and a lot to test for a one man project (like any of us could do, or like cordova-plugin-realm is), but would be fast enough for multiple persons that *do know how stuff works behind* when doing it.

in the extreme end the community would be in charge of doing this kind of feature, we would appreciate a sample android studio module and ios pod (with their usage guides), that we could later wrap in a cordova module.  
something like:
//run RealmStore.write( object : JsonObject) to open a write transaction,  given a valid RealmJS schema object.syncronous method. returns a boolean if success or not.
//run RealmStore.defineSchema(object : JsonObject,update : boolean) to define a schema. if already exists, it will be ignored unless update is sent. syncronous method. returns a boolean if success or not.

**Comment by bmunkholm** [comment]
@tryadelion As mentioned it's unfortunately not something that we can devote resources for at the moment. Depending on the help needed though, we would need to get some commitment from the community to this.
Perhaps @fealebenpae has input to how best to approach this.

**Comment by rethinkedu** [comment]
+1

**Comment by kabus202** [comment]
+1

**Comment by RoyGoode** [comment]
+1 for Ionic/Cordova

**Comment by clariontools** [comment]
+1 for Ionic/Cordova

**Comment by faytekin** [comment]
+1 for Cordova Plugin

**Comment by taxilian** [comment]
Hi; I have need of realm in either Cordova or Capacitor and would potentially be interested in taking responsibility for the maintenance and/or sponsoring some additional help with it; however, after spending about 6 hours digging through code to understand how things are set up I can see a number of different approaches which could be taken to try to make it work and I don't want to waste a lot of time implementing something that is counter to the goals of the rest of the project.

Is there an IRC, Slack, Discord, etc room where I could discuss with other developers to get some questions answered and make sure I'm on the same page with others?

**Comment by kneth** [comment]
@taxilian It is a very generous offer, thank you. I can fully understand if digging through the source code hasn't help you much.

We don't have a public chat forum but we can probably set up Slack channel if needed.

**Comment by taxilian** [comment]
That would be great. I am on the realm-public slack group and I will be working on it today. I’m starting to understand the JavaScript abstraction finally but there are definitely some areas that are a pain to track down. 

Richard

**Comment by SergioSuarezDev** [comment]
How is going with Realm plugin for Ionic? :D

**Comment by muriloferes** [comment]
Realm for Ionic! Please!

**Comment by ryu-man** [comment]
Is there any Idea how to use realm with ionic?

**Comment by saravanakumargn** [comment]
We should use `realm-web`?
https://www.mongodb.com/docs/realm/web/quickstart/

**Comment by taxilian** [comment]
realm-web requires an atlas web services backend. it's not helpful for
someone who just wants a standalone database without a subscription service.
